File: shared/ui/ipc/webview.protocol.commands.ts
import { FetchThirdPartyBoardsRequest } from "@codestream/protocols/agent";
import { CSMarker } from "@codestream/protocols/api";
import { RequestType } from "vscode-jsonrpc";

export interface GoToSlackSigninResponse {}

export const GoToSlackSigninRequestType = new RequestType<
	void,
	GoToSlackSigninResponse,
	void,
	void
>("extension/go-to-slack-signin");

export interface GetViewBootstrapDataResponse {
	unreads: any;
	// TODO: fill out the rest
}

export interface ValidateSignupRequest {
	token?: string;
}

export const ValidateSignupRequestType = new RequestType<
	ValidateSignupRequest,
	GetViewBootstrapDataResponse,
	void,
	void
>("extension/validate-signup");

export interface LoginRequest {
	email: string;
	password: string;
}
export const LoginRequestType = new RequestType<
	LoginRequest,
	GetViewBootstrapDataResponse,
	void,
	void
>("extension/authenticate");

export const GetViewBootstrapDataRequestType = new RequestType<
	void,
	GetViewBootstrapDataResponse,
	void,
	void
>("extension/bootstrap");

export interface ShowMarkersInEditorRequest {
	enable: boolean;
}

export interface ShowMarkersInEditorResponse {}

export const ShowMarkersInEditorRequestType = new RequestType<
	ShowMarkersInEditorRequest,
	ShowMarkersInEditorResponse,
	void,
	void
>("extension/show-markers");

export interface MuteAllConversationsRequest {
	mute: boolean;
}

export interface MuteAllConversationsResponse {}

export const MuteAllConversationsRequestType = new RequestType<
	MuteAllConversationsRequest,
	MuteAllConversationsResponse,
	void,
	void
>("extension/mute-all");

export interface OpenCommentOnSelectRequest {
	enable: boolean;
}

export interface OpenCommentOnSelectResponse {}

export const OpenCommentOnSelectInEditorRequestType = new RequestType<
	OpenCommentOnSelectRequest,
	OpenCommentOnSelectResponse,
	void,
	void
>("extension/open-comment-on-select");

export const getFetchIssueBoardsCommand = (service: string) =>
	new RequestType<FetchThirdPartyBoardsRequest, { boards: any[] }, void, void>(
		`codeStream/${service}/boards`
	);

export interface StartSignupRequest {}
export interface StartSignupResponse {}

export const StartSignupRequestType = new RequestType<
	StartSignupRequest,
	StartSignupResponse,
	void,
	void
>("extension/go-to-signup");

export interface SignOutRequest {}
export interface SignOutResponse {}

export const SignOutRequestType = new RequestType<SignOutRequest, SignOutResponse, void, void>(
	"extension/sign-out"
);

export interface ShowCodeRequest {
	marker: CSMarker;
	enteringThread: boolean;
}

export interface ShowCodeResponse {
	status: string;
}

export const ShowCodeRequestType = new RequestType<ShowCodeRequest, ShowCodeResponse, void, void>(
	"extension/show-code"
);

export interface HighlightCodeRequest {
	uri: string;
	marker: CSMarker;
	highlight: boolean;
	source: string;
}

export interface HighlightCodeResponse {
	result: any;
}

export const HighlightCodeRequestType = new RequestType<
	HighlightCodeRequest,
	HighlightCodeResponse,
	void,
	void
>("extension/highlight-code");

export interface HighlightLineRequest {
	uri: string;
	line: number;
	highlight: boolean;
	source: string;
}

export interface HighlightLineResponse {
	result: any;
}

export const HighlightLineRequestType = new RequestType<
	HighlightLineRequest,
	HighlightLineResponse,
	void,
	void
>("extension/highlight-line");

export interface RevealFileLineRequest {
	line: number;
}

export interface RevealFileLineResponse {}

export const RevealFileLineRequestType = new RequestType<
	RevealFileLineRequest,
	RevealFileLineResponse,
	void,
	void
>("extension/reveal-line");

export interface StartCommentOnLineRequest {
	line: number;
	uri: any;
	type: any;
}

export interface StartCommentOnLineResponse {}

export const StartCommentOnLineRequestType = new RequestType<
	StartCommentOnLineRequest,
	StartCommentOnLineResponse,
	void,
	void
>("extension/start-comment-on-line");

export const ReloadWebviewRequestType = new RequestType<void, void, void, void>(
	"extension/reload-webview"
);

export interface InviteToLiveShareRequest {
	userId: string;
	createNewStream: boolean;
}

export interface InviteToLiveShareResponse {}

export const InviteToLiveShareRequestType = new RequestType<
	InviteToLiveShareRequest,
	InviteToLiveShareResponse,
	void,
	void
>("extension/invite-to-liveshare");

export interface StartLiveShareRequest {
	streamId: string;
	threadId: string;
	createNewStream: boolean;
}

export interface StartLiveShareResponse {}

export const StartLiveShareRequestType = new RequestType<
	StartLiveShareRequest,
	StartLiveShareResponse,
	void,
	void
>("extension/start-liveshare");

export interface JoinLiveShareRequest {
	url: string;
}

export interface JoinLiveShareResponse {}

export const JoinLiveShareRequestType = new RequestType<
	JoinLiveShareRequest,
	JoinLiveShareResponse,
	void,
	void
>("extension/join-liveshare");

export interface ShowDiffRequest {
	marker: CSMarker;
}
export interface ShowDiffResponse {}

export const ShowDiffRequestType = new RequestType<ShowDiffRequest, ShowDiffResponse, void, void>(
	"extension/show-diff"
);

export interface ApplyPatchRequest {
	marker: CSMarker;
}

export interface ApplyPatchResponse {}

export const ApplyPatchRequestType = new RequestType<
	ApplyPatchRequest,
	ApplyPatchResponse,
	void,
	void
>("extension/apply-patch");

export interface UpdateConfigurationRequest {
	name: string;
	value: any;
}

export interface UpdateConfigurationResponse {}

export const UpdateConfigurationRequestType = new RequestType<
	UpdateConfigurationRequest,
	UpdateConfigurationResponse,
	void,
	void
>("extension/configuration/update");

File: vs/src/CodeStream.VisualStudio/Models/Webview.Protocol.Commands.cs
using System.Windows.Forms.VisualStyles;

namespace CodeStream.VisualStudio.Models
{
    public class EmptyRequestTypeParams {}

    //public static class Foo
    //{
    //    public static RequestType<EmptyRequestTypeParams> GoToSlackSigninRequestType = new RequestType<EmptyRequestTypeParams>("extension/go-to-slack-signin");
    //}

    public class GoToSlackSigninRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/go-to-slack-signin";
        public override string Method => MethodName;
    }

    public class StartSignupRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/go-to-signup";
        public override string Method => MethodName;
    }

    public class SignOutRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/sign-out";
        public override string Method => MethodName;
    }

    public class ValidateSignupRequest
    {
        public string Token { get; set; }
    }
    public class ValidateSignupRequestType : RequestType<ValidateSignupRequest>
    {
        public const string MethodName = "extension/validate-signup";
        public override string Method => MethodName;
    }

    public class LoginRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/authenticate";
        public override string Method => MethodName;
    }

    public class GetViewBootstrapDataRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/bootstrap";
        public override string Method => MethodName;
    }

    public class MuteAllConversationsRequestTypeParams
    {
        public bool Mute { get; set; }
    }
    public class MuteAllConversationsRequestType : RequestType<MuteAllConversationsRequestTypeParams>
    {
        public const string MethodName = "extension/mute-all";
        public override string Method => MethodName; 
    }

    public class ShowMarkersInEditorRequestTypeParams
    {
        public bool Enable { get; set; }
    }

    public class ShowMarkersInEditorRequestType : RequestType<ShowMarkersInEditorRequestTypeParams>
    {
        public const string MethodName = "extension/show-markers";
        public override string Method => MethodName;
    }

    public class OpenCommentOnSelectInEditorRequestTypeParams
    {
        public bool Enable { get; set; }
    }
    public class OpenCommentOnSelectInEditorRequestType : RequestType<OpenCommentOnSelectInEditorRequestTypeParams>
    {
        public const string MethodName = "extension/open-comment-on-select";
        public override string Method => MethodName;
    }

    public class ShowCodeRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/show-code";
        public override string Method => MethodName;
    }

    public class HighlightCodeRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/highlight-code";
        public override string Method => MethodName;
    }

    public class RevealFileLineRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/reveal-line";
        public override string Method => MethodName;
    }

    public class StartCommentOnLineRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/start-comment-on-line";
        public override string Method => MethodName;
    }

    public class ReloadWebviewRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/reload-webview";
        public override string Method => MethodName;
    }

    public class InviteToLiveShareRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/invite-to-liveshare";
        public override string Method => MethodName;
    }
    public class StartLiveShareRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/start-liveshare";
        public override string Method => MethodName;
    }

    public class JoinLiveShareRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/join-liveshare";
        public override string Method => MethodName;
    }
    public class ShowDiffRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/show-diff";
        public override string Method => MethodName;
    }

    public class ApplyPatchRequestType : RequestType<EmptyRequestTypeParams>
    {
        public const string MethodName = "extension/apply-patch";
        public override string Method => MethodName;
    }
}
